### PR TITLE
Replace load-generator with Go implementation

### LINKED ISSUE #688: Rewrite load-generator in Go
The `/tools/load-generator` is the only component of this repo that is in Python. It would be nice to replace this with a Go binary in order to eliminate the Python ecosystem dependency for one tool. There are also no CI tests for this Python code, so we don't know if it's broken when Python or other requirements change.

### LINKED ISSUE #705: PromBench Idea: exercise more parts of PromQL
PromBench runs synthetic load defined in [6_loadgen.yaml](https://github.com/prometheus/test-infra/blob/34a189da3778d9035d430e3213f5c9dd18d520a7/prombench/manifests/prombench/benchmark/6_loadgen.yaml).
These cover selectors, `rate`, `histogram_quantile` and `sum`, but leave out most of PromQL.

For instance we could add:
* binary arithmetic operators (joins)
* logical operators (`and`, `or`, `unless`)
* `topk`

### PR DESCRIPTION
**Pull request by HarshitNagpal2901**
Replace python load-generator with a Go implementation.
Add additional PromQL tests.
Fixes https://github.com/prometheus/test-infra/issues/688
Fixes https://github.com/prometheus/test-infra/issues/705

### CODE REVIEW

File: tools/load-generator/main.go [hunk 1]
```diff
@@ -0,0 +1,240 @@
+// Copyright 2019 The Prometheus Authors
```
Comment (SuperQ) [inline tools/load-generator/main.go]:
```suggestion
// Copyright 2024 The Prometheus Authors
```

File: tools/load-generator/main.go [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"gopkg.in/yaml.v2"
+)
+
+// Global variables and Prometheus metrics
+var (
+	max404Errors = 30
```
Comment (SuperQ) [inline tools/load-generator/main.go]: This should be a `const`.

**Review by SuperQ — CHANGES_REQUESTED** [review]

File: tools/load-generator/main.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+	}
+	prNumber := os.Args[2]
+
+	configFile, err := os.ReadFile("/etc/loadgen/config.yaml")
+	if err != nil {
+		log.Fatalf("Failed to load config: %v", err)
+	}
+	if err != nil {
+		log.Fatalf("Failed to load config: %v", err)
+	}
```
Comment (SuperQ) [inline tools/load-generator/main.go]: This appears to be duplicated code.

```suggestion
```

**Comment by HarshitNagpal2901** [comment]
@SuperQ sir, PTAL

File: tools/load-generator/main.go [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+	for i, group := range config.Querier.Groups {
+		wg.Add(1)
+		go NewQuerier(i, "pr", prNumber, group).run(&wg)
+	}
+
+	for i, group := range config.Querier.Groups {
+		wg.Add(1)
+		go NewQuerier(i, "release", prNumber, group).run(&wg)
+	}
```
Comment (SuperQ) [inline tools/load-generator/main.go]: This can be done in one loop.

```suggestion
	for i, group := range config.Querier.Groups {
		wg.Add(1)
		go NewQuerier(i, "pr", prNumber, group).run(&wg)
		wg.Add(1)
		go NewQuerier(i, "release", prNumber, group).run(&wg)
	}
```

**Comment by HarshitNagpal2901** [comment]
@SuperQ Sir, can u pls approve?

File: prombench/manifests/prombench/benchmark/6_loadgen.yaml [hunk 5]
```diff
@@ -48,6 +48,25 @@ data:
...
+        type: instant
+        queries:
+        - expr: rate(node_cpu_seconds_total{mode="system"}[5m]) and rate(node_cpu_seconds_total{mode="user"}[5m])
+        - expr: node_filesystem_avail_bytes unless node_filesystem_size_bytes
+        - expr: rate(node_network_receive_bytes_total[5m]) or rate(node_network_transmit_bytes_total[5m])
+      - name: topk_example
+        interval: 15s
+        type: instant
+        queries:
+        - expr: topk(5, rate(http_requests_total[5m]))
```
Comment (bboreham) [inline prombench/manifests/prombench/benchmark/6_loadgen.yaml:69]: Full marks for including it in the description, but this should be a separate PR.
Indeed it already is - #723.

File: tools/load-generator/Dockerfile [hunk 6]
```diff
@@ -1,10 +1,8 @@
-FROM python:3
+FROM quay.io/prometheus/busybox:latest
```
Comment (bboreham) [inline tools/load-generator/Dockerfile:1]: If possible we should use `FROM scratch`, to avoid giving attackers a shell.

File: tools/load-generator/main.go [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+	} else if resp.StatusCode != http.StatusOK {
+		log.Printf("WARNING: GroupID#%d: Querier returned %d for Prometheus instance %s.", q.groupID, resp.StatusCode, q.url)
+	} else {
+		body, _ := io.ReadAll(resp.Body)
+		log.Printf("GroupID#%d: query %s %s, status=%d, size=%d, duration=%.3f", q.groupID, q.target, expr, resp.StatusCode, len(body), duration.Seconds())
+	}
+}
+
+func durationSeconds(s string) time.Duration {
+	num := s[:len(s)-1]
```
Comment (bboreham) [inline tools/load-generator/main.go:187]: Crashes when s is empty.

**Review by bboreham — COMMENTED** [review]
Pretty good - crashed on first attempt but after a simple fix it seems to be running fine.

File: tools/load-generator/Dockerfile (hunk 6 above)
Comment (SuperQ) [reply tools/load-generator/Dockerfile:1]: The busybox image is our standard image. This is fine.

File: prombench/manifests/prombench/benchmark/6_loadgen.yaml (hunk 5 above)
Comment (HarshitNagpal2901) [reply prombench/manifests/prombench/benchmark/6_loadgen.yaml:69]: @bboreham Hi Sir, I am the same person who has created that #723 PR. I was not able to do DCO sign off for some weird reason, so i created a Fresh PR, from my secondary account, as i told that, my university exams are going on till september end, so as soon as it will end, i will look up to both the PRs. 
Hope you understand.❤️❤️

**Comment by bboreham** [comment]
I didn't want to push a commit to your `main` branch so created a new PR with my fix at #748.

**Comment by bboreham** [comment]
Replaced by #748
